test/RewardPool-test.js

```diff
@@ -240,144 +243,103 @@ contract("RewardPool", function (accounts) {
+        await rewardPool.submitPayeeMerkleRoot(rewardProgramID, 1, root, {
+          from: tally,
+        });
 
-        await rewardPool.submitPayeeMerkleRoot(updatedRoot, {
+        // This should not throw any errors
+        await rewardPool.submitPayeeMerkleRoot(rewardProgramID, 2, root, {
           from: tally,
         });
+      });
 
-        let paymentCycleNumber = await rewardPool.numPaymentCycles();
+      it("allows a multiple merkle roots for different reward programs to be submitted with the same payment cycle number", async function () {
+        let merkleTree = new PaymentTree(payments);
+        let root = merkleTree.getHexRoot();
+        await rewardPool.submitPayeeMerkleRoot(rewardProgramID, 1, root, {
+          from: tally,
+        });
 
-        assert.equal(
-          paymentCycleNumber.toNumber(),
-          previousPaymentCycleNumber.add(new BN(2)).toNumber(),
-          "the payment cycle number is correct"
-        );
+        // This should not throw any errors
+        await rewardPool.submitPayeeMerkleRoot(otherRewardProgramID, 1, root, {
+          from: tally,
+        });
       });
 
-      it("does not allow 2 merkle roots to be submitted in the same block after the previous payment cycle has ended", async function () {
+      it("does not allow a root to be submitted twice for the same reward program and cycle", async function () {
         let merkleTree = new PaymentTree(payments);
         let root = merkleTree.getHexRoot();
-        await rewardPool.submitPayeeMerkleRoot(root, { from: tally });
-
-        let updatedPayments = payments.slice();
-        updatedPayments[0].amount = updatedPayments[0].amount.add(
-          toTokenUnit(10)
-        );
-        let updatedMerkleTree = new PaymentTree(updatedPayments);
-        let updatedRoot = updatedMerkleTree.getHexRoot();
+        await rewardPool.submitPayeeMerkleRoot(rewardProgramID, 1, root, {
+          from: tally,
+        });
 
         await rewardPool
-          .submitPayeeMerkleRoot(updatedRoot, { from: tally })
+          .submitPayeeMerkleRoot(rewardProgramID, 1, root, {
+            from: tally,
+          })
           .should.be.rejectedWith(
             Error,
-            "Cannot start new payment cycle before currentPaymentCycleStartBlock"
+            "Payee root already submitted for this program & cycle"
           );
-        let paymentCycleNumber = await rewardPool.numPaymentCycles();
-
-        assert.equal(
-          paymentCycleNumber.toNumber(),
-          previousPaymentCycleNumber.add(new BN(1)).toNumber(),
-          "the payment cycle number is correct"
-        );
       });
 
       it("does not allow non-tally to submit merkle root", async function () {
         let merkleTree = new PaymentTree(payments);
         let root = merkleTree.getHexRoot();
 
         await rewardPool
-          .submitPayeeMerkleRoot(root, { from: accounts[11] })
+          .submitPayeeMerkleRoot(rewardProgramID, 1, root, {
```

Comment:
Yea. After looking at the test. It seems silly to rely on a paymentCycle specified by the caller to submit a merkle root. Shouldn't the smart contract know which paymentCycle you are wrong within its own state.

Reply:
So my idea is that each payment cycle is a fixed number of blocks, so that each miner / etc knows the range it's supposed to be calculating the rewards for (or at least how frequently to run) and they don't necessarily have to be run in order.

> Like what if tally loses track of the payment cycle. Is there a deeper reasoning behind this??

If you know the payment cycles you want calculated, you can see if they've already been calculated and submitted (or should be able to, I've realised the variable is private). Tally can keep track of which ones it's already done, but that is essentially an optimisation.

If we're catching the merkerootsubmission event in subgraph we would also have a record there of which ones had been done already.

We could make it sequential again (at least for now, and adjust later when we want to bring in decentralisation), the overall interface is very similar. In tally, we get the current payment cycle just before writing the root anyway, so could add it back into the contract code.

Reply:
> So my idea is that each payment cycle is a fixed number of blocks, so that each miner / etc knows the range it's supposed to be calculating the rewards for (or at least how frequently to run) and they don't necessarily have to be run in order.

Oh I see. Yes this is big change from previous implementation. Cool. **So computation can be out-of-sync with the write.** You can have cycle1, cycle3 already computed, with cycle2 still pending. **And writes don't have to be sequential.** But cycle3 can be written first, before cycle1. 

Correct?

> In tally, we get the current payment cycle just before writing the root anyway, so could add it back into the contract code.

So the choice here is between querying the payment cycle in the subgraph or on the contract. If we specify it in the contract, we will HAVE to assume sequentialism. So what I understand is that u intentionally don't put it in the contract, so tally can just query the last 4 root submissions, it might be (cycle1, ??, cycle3, cycle4). And then tally will decide if the computation for cycle2 is ready for write or if the computation for cycle5 is ready for write). So we can enforce sequentialism based on this decision by tally and information from the subgraph or we can just make it parallel. 

Correct?

Reply:
> Oh I see. Yes this is big change from previous implementation. Cool. So computation can be out-of-sync with the write. You can have cycle1, cycle3 already computed, with cycle2 still pending. And writes don't have to be sequential. But cycle3 can be written first, before cycle1.

Yes - that's right. While we're at this stage, we could make it sequential.

> So the choice here is between querying the payment cycle in the subgraph or on the contract. If we specify it in the contract, we will HAVE to assume sequentialism. So what I understand is that u intentionally don't put it in the contract, so tally can just query the last 4 root submissions, it might be (cycle1, ??, cycle3, cycle4). And then tally will decide if the computation for cycle2 is ready for write or if the computation for cycle5 is ready for write). So we can enforce sequentialism based on this decision by tally and information from the subgraph or we can just make it parallel.

> Correct?

Broadly, yes I think so.

I think this PR isn't complete actually - we need to either add things back in to support sequential work or also store the payment cycle length somewhere for each reward program.

Reply:
Or this could be in the rule blob...

Reply:
> I think this PR isn't complete actually - we need to either add things back in to support sequential work or also store the payment cycle length somewhere for each reward program.

Right this PR implementation seems to make things clear in a generalised sense. 

My first thought (in the big picture) is that in terms of logic, we don't need to enforce that merkle root writes occur one after another like we did before. If the computing node has incentive to compute the rewards, they will compute the rewards. If the writing node (or consensus game) has incentive, the root to the missing payment cycles should be written in. So perhaps there's no need to add anything anymore?

But this is so far down-the-line. Is this pr included in the audit?